src/reaction_lammps_mupt/reaction_template_builder/reaction_template_pipeline/map_reactant_atoms.py

```
@@ -1,249 +1,710 @@
+        r1, r2 = Chem.Mol(pair[0]), Chem.Mol(pair[1])
+        products = rxn.RunReactants((r1, r2))
+        
+        # Get substructure matches for each reactant against reaction templates
+        matches_1 = r1.GetSubstructMatches(rxn.GetReactants()[0])
+        matches_2 = r2.GetSubstructMatches(rxn.GetReactants()[1])
+
+        # Process each product set generated by the reaction
+        for product_set in products:
+            # Clean up any existing DataFrame
+            if "df" in locals():
+                del df
```

Ruff (0.14.14)

[error] Undefined name `df`

(F821)